compiler/rustc_parse/src/lib.rs

```diff
@@ -475,13 +512,23 @@ pub fn tokentree_probably_equal_for_proc_macro(
     first: &TokenTree,
     other: &TokenTree,
     sess: &ParseSess,
+    relaxed_delim_match: bool,
 ) -> bool {
     match (first, other) {
         (TokenTree::Token(token), TokenTree::Token(token2)) => {
             token_probably_equal_for_proc_macro(token, token2)
         }
         (TokenTree::Delimited(_, delim, tts), TokenTree::Delimited(_, delim2, tts2)) => {
-            delim == delim2 && tokenstream_probably_equal_for_proc_macro(&tts, &tts2, sess)
+            let delim_match = if relaxed_delim_match {
+                // `NoDelim` delimiters can appear in the captured tokenstream, but not
+                // in the reparsed tokenstream. Allow them to match with anything, so
+                // that we check if the two streams are structurally equivalent.
+                delim == delim2 || *delim == DelimToken::NoDelim || *delim2 == DelimToken::NoDelim
+            } else {
+                delim == delim2
+            };
+            delim_match
+                && tokenstream_probably_equal_for_proc_macro(&tts, &tts2, sess, relaxed_delim_match)
         }
         _ => false,
```

Comment:
Possible cases here are:
- `⟪ token_tree ⟫` -> `token_tree`
- `⟪ tokens ⟫` -> `( tokens )` (on the second reparse)

So, I think a more precise version of this function would be
```
pub fn tokentree_probably_equal_for_proc_macro(
    token: &TokenTree,
    reparsed_token: &TokenTree,
    sess: &ParseSess,
    relaxed_delim_match: bool,
) -> bool {
    match (token, reparsed_token) {
        (TokenTree::Token(token), TokenTree::Token(reparsed_token)) => {
            token_probably_equal_for_proc_macro(token, reparsed_token)
        }
        (
            TokenTree::Delimited(_, delim, tokens),
            TokenTree::Delimited(_, reparsed_delim, reparsed_tokens),
        ) if delim == reparsed_delim => {
            tokenstream_probably_equal_for_proc_macro(tokens, reparsed_tokens, sess)
        }
        (TokenTree::Delimited(_, DelimToken::NoDelim, tokens), reparsed_token) => {
            if relaxed_delim_match {
                if let TokenTree::Delimited(_, DelimToken::Paren, reparsed_tokens) = reparsed_token {
                    if tokenstream_probably_equal_for_proc_macro(tokens, reparsed_tokens, sess) {
                        return true;
                    }
                }
            }
            tokenstream_probably_equal_for_proc_macro(tokens, TokenStream::from(reparsed_token), sess)
        }
        _ => false,
    }
}
```

Reply:
Can avoid `TokenStream::from` by using something like
```rust
tokens.len() == 1 && token_probably_equal_for_proc_macro(tokens[0], reparsed_token)
```
though.